### PR TITLE
  Adapt privileges for DECLARE

### PR DESCRIPTION
**Pull request by mkleen**
Privieges for DECLARE are adapted that they are depending on
the privileges of the user for the related query.  

Fixes https://github.com/crate/crate/issues/13306

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/PrivilegesIntegrationTest.java [hunk 1]
```diff
@@ -733,4 +733,18 @@ public void testAccessesToPgConstraintEntriesWithRespectToPrivileges() throws Ex
...
         assertThat(newUserWithPrivilegesResult, is(superUserResult));
     }
+
+    @Test
+    public void test_declare_fetch_close_for_non_super_users() {
+        try (Session testUserSession = testUserSession()) {
+            execute("BEGIN;", null, testUserSession);
+            execute("DECLARE my_cursor NO SCROLL CURSOR FOR SELECT * FROM pg_catalog.pg_attribute order by attname", null, testUserSession);
+            assertThat(response.rowCount(), is(0L));
+            execute("FETCH 1 FROM my_cursor;", null, testUserSession);
```
Comment (mkleen) [inline server/src/test/java/io/crate/integrationtests/PrivilegesIntegrationTest.java:743]: `Fetch` and `Close` can only be tested once a cursor is declared, therefore it's tested in an integration test.

**Review by mfussenegger — COMMENTED** [review]
Wouldn't it be good enough to always allow fetch & close? If the user has permission to declare the statement, it's implied that they have permission to fetch & close. If the user doesn't have permission to declare the cursor, fetch & close will fail anyways because there is no cursor they could fetch or close on.

File: server/src/test/java/io/crate/integrationtests/PrivilegesIntegrationTest.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/PrivilegesIntegrationTest.java:743]: It's possible to use `declare` in a unit test. See `CursorTest` for examples.

**Comment by mkleen** [comment]
> Wouldn't it be good enough to always allow fetch & close? If the user has permission to declare the statement, it's implied that they have permission to fetch & close. If the user doesn't have permission to declare the cursor, fetch & close will fail anyways because there is no cursor they could fetch or close on.

Good point. It would also make the implementation simpler.

File: docs/appendices/release-notes/unreleased.rst [hunk 2]
```diff
@@ -104,6 +104,10 @@ Fixes
 .. stable branch. You can add a version label (`v/X.Y`) to the pull request for
 .. an automated mergify backport.
 
+- Fixed privileges for :ref:`DECLARE <sql-declare>`for cursors to make it
+  accessible for non-super-users. Privileges are adapted that they are depending 
+  on the privileges for the related query.
+
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst:110]:
```suggestion
- Fixed privileges for :ref:`DECLARE <sql-declare>`for cursors to make it
  accessible for non-super-users. All users can declare cursors as long as they
  have the required permissions for the query used within ``DECLARE``.
```

File: server/src/main/java/io/crate/action/sql/Cursor.java [hunk 3]
```diff
@@ -62,11 +64,13 @@ public Cursor(CircuitBreaker circuitBreaker,
                   boolean scroll,
                   Hold hold,
                   CompletableFuture<BatchIterator<Row>> queryIterator,
-                  List<Symbol> outputs) {
+                  List<Symbol> outputs,
+                  AnalyzedStatement query) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/action/sql/Cursor.java:68]: Looks like this change is no longer necessary?

The privilege check uses the query on `AnalyzedDeclare`, not on the `Cursor` ?

Comment (mkleen) [reply server/src/main/java/io/crate/action/sql/Cursor.java]: Yes, you are right. Thank you!

**Review by matriv — APPROVED** [review]
LGTM, please wait for Matthias to approve as well, as he was the original reviewer.

**Review by mfussenegger — COMMENTED** [review]
Could you add tests for fetch and close as well?

```
diff --git a/server/src/test/java/io/crate/auth/AccessControlMayExecuteTest.java b/server/src/test/java/io/crate/auth/AccessControlMayExecuteTest.java
index 037e2cea2f21..19d48320298c 100644
--- a/server/src/test/java/io/crate/auth/AccessControlMayExecuteTest.java
+++ b/server/src/test/java/io/crate/auth/AccessControlMayExecuteTest.java
@@ -26,11 +26,10 @@
 import static java.util.Collections.singletonList;
 import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.hasItem;
 import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertThat;
-import static org.junit.Assert.assertThrows;
 import static org.mockito.Mockito.mock;
 
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
@@ -47,22 +46,29 @@
 import org.elasticsearch.test.ClusterServiceUtils;
 import org.hamcrest.Matcher;
 import org.hamcrest.Matchers;
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.Answers;
 
 import io.crate.analyze.ParamTypeHints;
 import io.crate.analyze.TableDefinitions;
+import io.crate.data.Row;
 import io.crate.exceptions.UnauthorizedException;
 import io.crate.execution.engine.collect.sources.SysTableRegistry;
 import io.crate.metadata.RelationName;
 import io.crate.metadata.cluster.DDLClusterStateService;
 import io.crate.metadata.settings.CoordinatorSessionSettings;
+import io.crate.planner.DependencyCarrier;
+import io.crate.planner.Plan;
+import io.crate.planner.operators.SubQueryResults;
+import io.crate.protocols.postgres.TransactionState;
 import io.crate.sql.parser.SqlParser;
 import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
 import io.crate.testing.SQLExecutor;
 import io.crate.testing.T3;
+import io.crate.testing.TestingRowConsumer;
 import io.crate.user.Privilege;
 import io.crate.user.User;
 import io.crate.user.UserLookupService;
 import io.crate.user.UserManager;
 import io.crate.user.UserManagerService;
@@ -138,10 +144,22 @@ public User findUser(String userName) {
             .addView(new RelationName("doc", "v1"), "select * from users")
             .setUserManager(userManager)
             .build();
     }
 
+    private TestingRowConsumer executePlan(Plan plan) {
+        TestingRowConsumer consumer = new TestingRowConsumer(true);
+        plan.execute(
+            mock(DependencyCarrier.class, Answers.RETURNS_MOCKS),
+            e.getPlannerContext(clusterService.state()),
+            consumer,
+            Row.EMPTY,
+            SubQueryResults.EMPTY
+        );
+        return consumer;
+    }
+
     private void analyze(String stmt) {
         analyze(stmt, user);
     }
 
     private void analyzeAsSuperUser(String stmt) {
@@ -706,6 +724,21 @@ public void test_anaylze_works_for_normal_user_with_AL_privileges() {
     public void test_declare_cursor_for_non_super_users() {
         analyze("DECLARE this_cursor NO SCROLL CURSOR FOR SELECT * FROM sys.summits;", user);
         assertAskedForTable(Privilege.Type.DQL, "sys.summits");
     }
 
+    @Test
+    public void test_fetch_from_cursor_is_allowed_if_user_could_declare_it() throws Exception {
+        Plan plan = e.plan("declare c1 no scroll cursor for select 1");
+        e.transactionState= TransactionState.IN_TRANSACTION;
+        executePlan(plan);
+        analyze("fetch from c1");
+    }
+
+    @Test
+    public void test_close_cursor_is_allowed_if_user_could_declare_it() throws Exception {
+        Plan plan = e.plan("declare c1 no scroll cursor for select 1");
+        e.transactionState= TransactionState.IN_TRANSACTION;
+        executePlan(plan);
+        analyze("close c1");
+    }
 }
 ```

**Comment by mkleen** [comment]
> Could you add tests for fetch and close as well?

sure, thanks!

**Review by mfussenegger — APPROVED** [review]
Sphinx and checkstyle are unhappy - otherwise lgtm

**Pull request merged**